File: sdk/java-sdk-protobuf/src/main/java/kalix/javasdk/StatusCode.java
/*
 * Copyright (C) 2021-2024 Lightbend Inc. <https://www.lightbend.com>
 */

package kalix.javasdk;

/**
 * Interface used to represent HTTP status code. **NOT** for user extension.
 */
public interface StatusCode {

  // return the value of the status code
  int value();

  enum Success implements StatusCode {
    OK(200),
    CREATED(201),
    ACCEPTED(202),
    NON_AUTHORITATIVE_INFORMATION(203),
    NO_CONTENT(204),
    RESET_CONTENT(205),
    PARTIAL_CONTENT(206),
    MULTI_STATUS(207),
    ALREADY_REPORTED(208),
    IM_USED(226);

    private final int value;

    Success(int value) {
      this.value = value;
    }

    @Override
    public int value() {
      return value;
    }

    public static Success from(int statusCode) {
      for (Success s : Success.values()) {
        if (s.value() == statusCode) {
          return s;
        }
      }
      throw new IllegalArgumentException("No matching status for [" + statusCode + "]");
    }
  }

  enum Redirect implements StatusCode {
    MULTIPLE_CHOICES(300),
    MOVED_PERMANENTLY(301),
    FOUND(302),
    SEE_OTHER(303),
    NOT_MODIFIED(304),
    USE_PROXY(305),
    TEMPORARY_REDIRECT(307),
    PERMANENT_REDIRECT(308);

    private final int value;

    Redirect(int value) {
      this.value = value;
    }

    @Override
    public int value() {
      return value;
    }
  }


  /** The supported HTTP error codes that can be used when replying from the Kalix user function. */
  enum ErrorCode implements StatusCode {
    BAD_REQUEST(400),
    UNAUTHORIZED(401),
    FORBIDDEN(403),
    NOT_FOUND(404),
    CONFLICT(409),
    TOO_MANY_REQUESTS(429),
    INTERNAL_SERVER_ERROR(500),
    SERVICE_UNAVAILABLE(503),
    GATEWAY_TIMEOUT(504);

    private final int value;

    ErrorCode(int value) {
      this.value = value;
    }

    @Override
    public int value() {
      return value;
    }
  }
}

File: sdk/scala-sdk-protobuf/src/main/scala/kalix/scalasdk/StatusCode.scala
/*
 * Copyright (C) 2021-2024 Lightbend Inc. <https://www.lightbend.com>
 */

package kalix.scalasdk

/**
 * A sealed trait representing HTTP status codes.
 */
sealed trait StatusCode

/**
 * Companion object for the `StatusCode` trait, containing various HTTP status code definitions.
 */
object StatusCode {
  sealed abstract class Success(val value: Int) extends StatusCode

  case object Ok extends Success(200)

  case object Created extends Success(201)

  case object Accepted extends Success(202)

  case object NonAuthoritativeInformation extends Success(203)

  case object NoContent extends Success(204)

  case object ResetContent extends Success(205)

  case object PartialContent extends Success(206)

  case object MultiStatus extends Success(207)

  case object AlreadyReported extends Success(208)

  case object IMUsed extends Success(226)

  sealed abstract class Redirect(val value: Int) extends StatusCode

  case object MultipleChoices extends Redirect(300)

  case object MovedPermanently extends Redirect(301)

  case object Found extends Redirect(302)

  case object SeeOther extends Redirect(303)

  case object NotModified extends Redirect(304)

  case object UseProxy extends Redirect(305)

  case object TemporaryRedirect extends Redirect(307)

  case object PermanentRedirect extends Redirect(308)

  sealed abstract class Error(val value: Int) extends StatusCode

  case object BadRequest extends Error(400)

  case object Unauthorized extends Error(401)

  case object Forbidden extends Error(403)

  case object NotFound extends Error(404)

  case object Conflict extends Error(409)

  case object TooManyRequests extends Error(429)

  case object InternalServerError extends Error(500)

  case object ServiceUnavailable extends Error(503)

  case object GatewayTimeout extends Error(504)

}
